File: python/templative/lib/distribute/playground/deeper/playgroundTextureMaker.py
from os import path
from PIL import Image
import math
from templative.lib.componentInfo import COMPONENT_INFO

async def createCompositeImageInTextures(componentName, componentTypeInfo, frontInstructions, textureDirectoryFilepath):
    totalCount = 0
    for instruction in frontInstructions:
        totalCount += instruction["quantity"]

    if totalCount == 0:
        return 0,0,0
    
    pixelDimensions = componentTypeInfo["DimensionsPixels"]
    columns = math.floor(math.sqrt(totalCount))
    rows = columns
    while columns * rows < totalCount:
        rows += 1

    tiledImage = Image.new('RGB',(pixelDimensions[0]*columns, pixelDimensions[1]*rows))
    
    xIndex = 0
    yIndex = 0
    for instruction in frontInstructions:
        image = Image.open(instruction["filepath"])
        for _ in range(instruction["quantity"]):
            tiledImage.paste(image,(xIndex*pixelDimensions[0],yIndex*pixelDimensions[1]))
            xIndex += 1
            if xIndex == columns:
                xIndex = 0
                yIndex +=1

    frontImageName = "%s-front.jpeg" % componentName
    frontImageFilepath = path.join(textureDirectoryFilepath, frontImageName)
    tiledImage.save(frontImageFilepath,"JPEG")
    return totalCount, columns, rows

File: src/main/templative/lib/distribute/playground/deeper/playgroundTextureMaker.js
const path = require('path');
const { Image } = require('image-js');
const { COMPONENT_INFO } = require('../../../componentInfo.js');

async function createCompositeImageInTextures(componentName, componentTypeInfo, frontInstructions, textureDirectoryFilepath) {
  let totalCount = 0;
  for (const instruction of frontInstructions) {
    totalCount += instruction.quantity;
  }

  if (totalCount === 0) {
    return [0, 0, 0];
  }

  const pixelDimensions = componentTypeInfo.DimensionsPixels;
  let columns = Math.floor(Math.sqrt(totalCount));
  let rows = columns;
  while (columns * rows < totalCount) {
    rows += 1;
  }

  const tiledImage = new Image({
    width: pixelDimensions[0] * columns,
    height: pixelDimensions[1] * rows,
    colorModel: 'rgb'
  });

  let xIndex = 0;
  let yIndex = 0;
  for (const instruction of frontInstructions) {
    const image = await Image.load(instruction.filepath);
    for (let i = 0; i < instruction.quantity; i++) {
      tiledImage.drawImage(image, xIndex * pixelDimensions[0], yIndex * pixelDimensions[1]);
      xIndex += 1;
      if (xIndex === columns) {
        xIndex = 0;
        yIndex += 1;
      }
    }
  }

  const frontImageName = `${componentName}-front.jpeg`;
  const frontImageFilepath = path.join(textureDirectoryFilepath, frontImageName);
  await tiledImage.save(frontImageFilepath, { format: 'jpeg' });
  
  return [totalCount, columns, rows];
}

module.exports = {
  createCompositeImageInTextures
};
